python/ray/util/spark/cluster_init.py

```diff
@@ -0,0 +1,758 @@
+        except Exception:
+            os.close(fd)
+
+    lock_file_path = "/tmp/ray_on_spark_worker_startup_barrier_lock.lock"
+    try:
+        lock_fd = acquire_lock(lock_file_path)
+    except TimeoutError:
+        # If timeout happens, the file lock might be hold by another process and that process
+        # does not release the lock in time by some unexpected reason.
+        # In this case, remove the existing lock file and create the file again, and then
+        # acquire file lock on the new file.
+        try:
+            os.remove(lock_file_path)
+        except Exception:
+            pass
+        lock_fd = acquire_lock(lock_file_path)
+
+    def release_lock():
+        fcntl.flock(lock_fd, fcntl.LOCK_UN)
+        os.close(lock_fd)
+
+    try:
+        port_alloc_file = "/tmp/ray_on_spark_worker_port_allocation.txt"
+
+        # NB: reading / writing `port_alloc_file` is protected by exclusive lock
+        # on file `lock_file_path`
+        if os.path.exists(port_alloc_file):
+            with open(port_alloc_file, mode="r") as fp:
+                port_alloc_data = fp.read()
+            port_alloc_table = [
+                line.split(" ") for line in port_alloc_data.strip().split("\n")
+            ]
+            port_alloc_table = [
+                (int(pid_str), int(slot_index_str))
+                for pid_str, slot_index_str in port_alloc_table
+            ]
+        else:
+            port_alloc_table = []
+            with open(port_alloc_file, mode="w"):
+                pass
+            # The port range allocation file must be readable / writable to all users.
+            os.chmod(port_alloc_file, 0o0777)
+
+        port_alloc_map = {
+            pid: slot_index
+            for pid, slot_index in port_alloc_table
+            if psutil.pid_exists(pid)  # remove slot used by dead process
+        }
+
+        allocated_slot_set = set(port_alloc_map.values())
+
+        if len(allocated_slot_set) == 0:
+            new_slot_index = 0
+        else:
+            new_slot_index = max(allocated_slot_set) + 1
+            for index in range(new_slot_index):
+                if index not in allocated_slot_set:
+                    new_slot_index = index
+                    break
+
+        port_alloc_map[os.getpid()] = new_slot_index
+
+        with open(port_alloc_file, mode="w") as fp:
+            for pid, slot_index in port_alloc_map.items():
+                fp.write(f"{pid} {slot_index}\n")
+
+        worker_port_range_begin = 20000 + new_slot_index * 1000
+        worker_port_range_end = worker_port_range_begin + 1000
+
+        if worker_port_range_end > 65536:
+            raise RuntimeError(
+                "Too many ray worker nodes are running on this machine, cannot "
+                "allocate worker port range for new ray worker node."
+            )
+    except Exception:
+        release_lock()
+        raise
+
+    def hold_lock_for_10s_and_release():
```

Comment:
In ray spark routine, it cannot get which ports are allocated and used by ray nodes, so 10s is just an estimation, after 10s, the ray nodes should be started and bind allocated ports. So the logic is following:

Inside a spark task:
1. acquire exclusive lock
2. create a thread to run "hold_lock_for_10s_and_release"
3. start ray node in blocking way, this line code blocks forever, but after 10s, all allocated ports should be bind by the started ray node

So that the first 10 seconds of "start ray node" is protected by the exclusive lock, it ensures the ray nodes can detect the unused port and then bind to the port safely.